Enable `outline-atomics` by default on more AArch64 platforms

tgross35 opened the pull request:
The baseline Armv8.0 ISA doesn't have atomics instructions, but in
practice most hardware is at least Armv8.1-A (2014), which includes
single-instruction atomics as part of the LSE feature. As a performance
optimization for these cases, GCC and LLVM have the `-moutline-atomics` flag
to turn atomic operations into calls to symbols like `__aarch64_cas1_acq`.
These can do runtime feature detection and use the LSE instructions if
available, falling back to more portable load-exclusive/store-exclusive
loops.

Since the recent 3b50253b57b1 ("compiler-builtins: plumb LSE support
for aarch64 on linux") our builtins support this LSE optimization, and
since 6936bb975a50 ("Dynamically enable LSE for aarch64 rust provided
intrinsics"), std will set the flag as part of its startup code. The first
commit in this PR configures this to work on all platforms built with
`outline-atomics`, not just Linux.

Thus, enable `outline-atomics` by default on Android, OpenBSD, Windows,
and Fuchsia platforms that don't have LSE in the baseline. The feature is
already enabled on Linux. Platform-specific details are included in each
commit message.

The current implementation can still be accessed by setting
`-Ctarget-feature=-outline-atomics`. Setting `-Ctarget-feature=+lse` or
a relevant CPU will use the single-instruction atomics without the call
overhead. https://rust.godbolt.org/z/dsdrzszoe

Link: https://learn.arm.com/learning-paths/servers-and-cloud-computing/lse/intro/
Original Clang outline-atomics benchmarks: https://reviews.llvm.org/D91157#2435844

try-job: aarch64-msvc-*
try-job: arm-android
try-job: dist-android
try-job: dist-aarch64-llvm-mingw
try-job: dist-aarch64-msvc
try-job: dist-various-*
try-job: test-various

tgross35 committed:
compiler-builtins: Change gating for outline atomic symbols

Build outline atomic symbols on all targets that have `outline-atomics`
enabled, rather than only on Linux. Since this is no longer OS-specific,
also rename the module.

tgross35 committed:
Make the `RUST_LSE_INIT` constructor cross-platform

Change the gating and link sections to enable this for any platforms
that enable `outline-atomics`, rather than only Linux. Additionally, no
longer run this if LSE is available, since in this case the outline
versions will never be called.

bors commented:
:umbrella: The latest upstream changes (presumably #145074) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 committed:
Enable `outline-atomics` by default on AArch64 Windows platforms

Windows has a similar flag `/forceInterlockedFunctions`, which uses
names such as `_InterlockedAdd64_rel`.

tgross35 committed:
Enable `outline-atomics` by default on AArch64 Android

Per LLVM commit c5e7e649d537 ("[AArch64][Clang][Linux] Enable
out-of-line atomics by default.") [1], Clang enables these on Android.
Thus, do the same in Rust.

[1]: https://github.com/llvm/llvm-project/commit/c5e7e649d537067de

tgross35 committed:
Enable `outline-atomics` by default on AArch64 Fuchsia

Clang has done this by default since LLVM commit 1a963d3278c2 ("[Driver]
Make -moutline-atomics default for aarch64-fuchsia targets"), [1], so do
the same here.

[1]: https://github.com/llvm/llvm-project/commit/1a963d3278c2daab7e12125371442cd129c09954

tgross35 committed:
Enable `outline-atomics` by default on AArch64 OpenBSD

Clang has recently started doing this, as of LLVM commit 5d774ec8d183
("[Driver] Enable outline atomics for OpenBSD/aarch64") [1]. Thus, do
the same here.

[1]: https://github.com/llvm/llvm-project/commit/5d774ec8d183cbbb243c57f50d891822211d3ec2

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17502885604/job/49719762662) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49719762662) [(plain)](https://github.com/rust-lang/rust/commit/effc5ae2edda3c8966b38d7e4882719f98027d59/checks/49719762662/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling ureq v3.0.8
   Compiling citool v0.1.0 (/home/runner/work/rust/rust/src/ci/citool)
    Finished `dev` profile [unoptimized] target(s) in 20.25s
     Running `target/debug/citool calculate-job-matrix`
Run type: TryJob { job_patterns: Some(["aarch64-apple", "aarch64-msvc-1", "aarch64-msvc-2", "arm-android", "dist-android", "dist-aarch64-windows-gnullvm", "dist-aarch64-apple", "dist-aarch64-msvc", "dist-various", "dist-x86_64-freebsd"]) }
Error: Failed to calculate job matrix

Caused by:
    Patterns `dist-various` did not match any auto jobs
##[error]Process completed with exit code 1.
```

</details>

madsmtm reviewed (commented):
Looked through it as best I could, but this is a fair bit outside my comfort zone.

Is there a test or two I could run to verify whether this works on-device?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/144938/6c699a37235700ab749e3f14147fe41d49c056e8..4c4f34f82cbd20fa51cfc0f446a371767726bd58)*

semarie commented:
@tgross35 thanks for the Cc for OpenBSD.

I checked with people involved in low level aarch64, and it should be fine for OpenBSD.

We are already using `-moutline-atomics` for some time in the kernel and we recently made it the default for our base clang (based on llvm-19). We also have a libcompiler-rt with runtime feature detection code necessary to enable LSE on systems that support it (which was upstreamed).

davidtwco reviewed (approved):
LGTM, up to you if you want to wait to hear back from the target maintainers individually before r=me

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/144938/397f93362974d298b79e0e0cd43677014aa7b722..b18a6c6817458b127f9f341484277687f999e82a)*

Nashenas88 commented:
I think this is good for fuchsia. I'm going to run it by some folks tomorrow morning.

Nashenas88 commented:
No concerns from other folks on fuchsia. Thanks for working on this!

Nashenas88 commented on library/std/src/sys/configure_builtins.rs:
```diff
@@ -1,8 +1,35 @@
+)]
 static RUST_LSE_INIT: extern "C" fn() = {
     extern "C" fn init_lse() {
         use crate::arch;
```
This comment is actually for line 37 below. Shouldn't the comment be updated to refer to the renamed file?
```
-// This is provided by compiler-builtins::aarch64_linux.
+// This is provided by compiler-builtins::aarch64_outline_atomics.
```

madsmtm commented on library/std/src/sys/configure_builtins.rs:
```diff
@@ -1,6 +1,10 @@
-#[cfg(all(target_arch = "aarch64", target_os = "linux", not(feature = "compiler-builtins-c")))]
+#[cfg(all(
+    target_arch = "aarch64",
+    target_feature = "outline-atomics",
```
I think I understand the feature a bit better now, and I'd be fine with removing the gate btw. But it's probably fine to leave it as-is for now, `+outline-atomics` doesn't work anyhow if you don't build `compiler-builtins` with that enabled as well.

madsmtm reviewed (commented):
I verified that `__rust_enable_lse` is run on Apple platforms when having `-lse,+outline-atomics`.

It isn't actually _used_ though, the code will always fall back to the slower, non LSE implementation (regardless of whether the current CPU actually supports it), since `try_lse_op!` is gated with `.arch_extension lse`, which doesn't seem to be enabled on Apple platforms?

And if I just remove it, I get:

```
error: ADR/ADRP relocations must be GOT relative
  |
note: instantiated into assembly here
 --> <inline asm>:6:14
  |
6 | adrp    x16, __ZN17compiler_builtins23aarch64_outline_atomics16HAVE_LSE_ATOMICS17h7f004a93068845c5E; nop; ldrb    w16, [x16, :lo12:__ZN17compiler_builtins...
  |              ^

error: unknown AArch64 fixup kind!
  |
note: instantiated into assembly here
 --> <inline asm>:6:14
  |
6 | adrp    x16, __ZN17compiler_builtins23aarch64_outline_atomics16HAVE_LSE_ATOMICS17h7f004a93068845c5E; nop; ldrb    w16, [x16, :lo12:__ZN17compiler_builtins...
  |              ^
```

That error seems to come from [around here in `AArch64MachObjectWriter.cpp`](https://github.com/llvm/llvm-project/blob/21857ae337e0892a5522b6e7337899caa61de2a6/llvm/lib/Target/AArch64/MCTargetDesc/AArch64MachObjectWriter.cpp#L194), which seems to hint that the way the check is implemented isn't supported on Mach-O? Unsure.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/144938/397f93362974d298b79e0e0cd43677014aa7b722..b18a6c6817458b127f9f341484277687f999e82a)*

madsmtm commented on library/std/src/sys/configure_builtins.rs:
```diff
@@ -1,8 +1,35 @@
-/// Hook into .init_array to enable LSE atomic operations at startup, if
-/// supported.
-#[cfg(all(target_arch = "aarch64", target_os = "linux", not(feature = "compiler-builtins-c")))]
+/// Enable LSE atomic operations at startup, if supported.
+///
+/// Linker sections are based on what [`ctor`] does, with priorities to run slightly before user
+/// code:
+///
+/// - Apple uses the section `__mod_init_func`, `mod_init_funcs` is needed to set
+///   `S_MOD_INIT_FUNC_POINTERS`. There doesn't seem to be a way to indicate priorities.
+/// - Windows uses `.CRT$XCT`, which is run before user constructors (these should use `.CRT$XCU`).
+/// - ELF uses `.init_array` with a priority of 90, which runs before our `ARGV_INIT_ARRAY`
+///   initializer (priority 99). Both are within the 0-100 implementation-reserved range, per docs
+///   for the [`prio-ctor-dtor`] warning, and this matches compiler-rt's `CONSTRUCTOR_PRIORITY`.
+///
+/// To save startup time, the initializer is only run if outline atomic routines from
+/// compiler-builtins may be used. If LSE is known to be available then the calls are never
+/// emitted, and if we build the C intrinsics then it has its own initializer using the symbol
+/// `__aarch64_have_lse_atomics`.
+///
+/// [`ctor`]: https://github.com/mmastrac/rust-ctor/blob/63382b833ddcbfb8b064f4e86bfa1ed4026ff356/shared/src/macros/mod.rs#L522-L534
+/// [`prio-ctor-dtor`]: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
```
Reflecting on it, I think it'd be useful with a comment here why we do this in a static initialiser, and aren't doing it in `std`'s `init` function in `rt.rs`?

I can think of an answer myself: because this affects atomics, which are in `core`, and because it'd be confusing if `#![no_main]` negatively affected atomics performance - but in that case, why aren't we doing the initialisation in core to start with, to not affect `#![no_std]`? The answer here is probably because we need runtime feature detection, which is only available in `std`.

But again, it'd be nice to have a discussion about this in the docs.

davidtwco replied:
I think this would make sense as an addition.

tgross35 replied:
I didn't know of any strong reason for `init_array` other than following precedent; C compiler-rt does a global constructor as well, though doing it before main isn't really an option with how source is organized there. Our comment on `ARGV_INIT_ARRAY` has a good point though, that using `init_array` means things work even if we're in a dylib, or `no_main` (as you mentioned).

Feature detection is indeed the reason to be in std as aarch64 needs auxvec. `compiler-rt` [can do that](https://github.com/llvm/llvm-project/blob/e66e8aac6e4c0cc3874cabad736233440b0ffb1e/compiler-rt/lib/builtins/cpu_model/aarch64.c#L37-L54) but it would be a mess trying to depend on `std` or `libc` symbols from compiler-builtins.

I added a comment to this effect.

tgross35 commented:
@bors2 try

tgross35 commented:
I think we could just enable it on all aarch64 targets by default at this point (Apple as a temporary exclusion for https://github.com/rust-lang/rust/issues/149634), opened https://github.com/rust-lang/rust/issues/150455 for that. Or doing more piecemeal is fine if anyone is interested in putting a PR up (I'm a bit out of time).

madsmtm commented:
> If I understand correctly, this is adding a static initializer to any code linking std on (most) aarch64 targets.

It was partly pre-existing, but yes. See [the comment in this file](https://github.com/tgross35/rust/blob/08f833aa179952037d5d0c5d949d4c3ec18ec212/library/std/src/sys/configure_builtins.rs#L22-L29) for reasoning behind it.

> Isn't this a problem for embedded use cases as well?

This is only done for performance, correctness is not affected if this ends up not running.

madsmtm commented:
> Some of our developers are hitting SIGILL / ILL_ILLOPC crashes which bisects to this change, so it would be really nice to have a knob to turn it off without spamming the logs.

Damn. Could you share more details here? Perhaps we should consider reverting this if it's a big issue?

> I'm confused by the `unstable feature` warning, because it seems to be saying `outline-atomics` is unstable, but we're actually trying to turn it off (`-Ctarget-feature=-outline-atomics`) so why is it warning about that?

The name "`outline-atomics`" itself is unstable too, and since it's unstable, it might be still be removed one day - and then the `-Ctarget-feature=-outline-atomics` would break.
